src/PowerShellEditorServices/Session/InvocationEventQueue.cs

```diff
@@ -0,0 +1,251 @@
+using System;
+using System.Collections.Generic;
+using System.Management.Automation.Runspaces;
+using System.Reflection;
+using System.Text;
+using System.Threading.Tasks;
+using System.Threading;
+
+namespace Microsoft.PowerShell.EditorServices.Session
+{
+    using System.Management.Automation;
+
+    /// <summary>
+    /// Provides the ability to take over the current pipeline in a runspace.
+    /// </summary>
+    internal class InvocationEventQueue
+    {
+        private readonly PromptNest _promptNest;
+
+        private readonly Runspace _runspace;
+
+        private readonly PowerShellContext _powerShellContext;
+
+        private InvocationRequest _invocationRequest;
+
+        private SemaphoreSlim _lock = new SemaphoreSlim(1, 1);
+
+        internal InvocationEventQueue(PowerShellContext powerShellContext, PromptNest promptNest)
+        {
+            _promptNest = promptNest;
+            _powerShellContext = powerShellContext;
+            _runspace = powerShellContext.CurrentRunspace.Runspace;
+            CreateInvocationSubscriber();
+        }
+
+        /// <summary>
+        /// Executes a command on the main pipeline thread through
+        /// eventing. A <see cref="PSEngineEvent.OnIdle" /> event subscriber will
+        /// be created that creates a nested PowerShell instance for
+        /// <see cref="PowerShellContext.ExecuteCommand" /> to utilize.
+        /// </summary>
+        /// <remarks>
+        /// Avoid using this method directly if possible.
+        /// <see cref="PowerShellContext.ExecuteCommand" /> will route commands
+        /// through this method if required.
+        /// </remarks>
+        /// <typeparam name="TResult">The expected result type.</typeparam>
+        /// <param name="psCommand">The <see cref="PSCommand" /> to be executed.</param>
+        /// <param name="errorMessages">
+        /// Error messages from PowerShell will be written to the <see cref="StringBuilder" />.
+        /// </param>
+        /// <param name="executionOptions">Specifies options to be used when executing this command.</param>
+        /// <returns>
+        /// An awaitable <see cref="Task" /> which will provide results once the command
+        /// execution completes.
+        /// </returns>
+        internal async Task<IEnumerable<TResult>> ExecuteCommandOnIdle<TResult>(
+            PSCommand psCommand,
+            StringBuilder errorMessages,
+            ExecutionOptions executionOptions)
+        {
+            var request = new PipelineExecutionRequest<TResult>(
+                _powerShellContext,
+                psCommand,
+                errorMessages,
+                executionOptions);
+
+            await SetInvocationRequestAsync(
+                new InvocationRequest(
+                    pwsh => request.Execute().GetAwaiter().GetResult()));
```

Comment:
`pwsh` would indicate it is only PSCore, would the more generic `powershell` be better? Same in other places as well

Reply:
So there's actually a specific reason why I have a habit of using `pwsh`, and that's entirely so method chaining lines up nicely.

```csharp
using (var pwsh = PowerShell.Create())
{
     pwsh.AddCommand("Get-ChildItem")
         .AddParameter("Path", "C:\\")
         .Invoke();
}
```

Personally I think it ends up being more readable, and `powershell` kind of has the opposite problem imo.  But obviously that's just a style choice so if others agree I don't have any issues changing it :P